velox/adapters/s3fs/S3FileSystem.cpp

```diff
@@ -0,0 +1,267 @@
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+
+#include <aws/core/Aws.h>
+#include <aws/core/Region.h>
+#include <aws/core/auth/AWSCredentialsProviderChain.h>
+#include <aws/core/http/HttpResponse.h>
+#include <aws/core/utils/logging/ConsoleLogSystem.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+
+namespace facebook::velox {
+
+constexpr std::string_view kSep{"/"};
+constexpr std::string_view s3Scheme{"s3:"};
+
+bool inline isS3File(const std::string_view filename) {
+  return (filename.substr(0, s3Scheme.size()) == s3Scheme);
+}
+
+inline Aws::String getAwsString(const std::string& s) {
+  return Aws::String(s.begin(), s.end());
+}
+
+inline std::string_view fromAwsString(const Aws::String& s) {
+  return {s.data(), s.length()};
+}
+
+// Implement the S3ReadFile
+class S3ReadFile final : public ReadFile {
+ public:
+  S3ReadFile(std::string path, std::shared_ptr<Aws::S3::S3Client> client)
+      : path_(std::move(path)), client_(client) {}
+
+  std::string_view pread(uint64_t offset, uint64_t length, Arena* arena)
+      const override {
+    char* pos = arena->reserve(length);
+    preadInternal(offset, length, pos);
+    return {pos, length};
+  }
+
+  std::string_view pread(uint64_t offset, uint64_t length, void* buf)
+      const override {
+    preadInternal(offset, length, static_cast<char*>(buf));
+    return {static_cast<char*>(buf), length};
+  }
+
+  std::string pread(uint64_t offset, uint64_t length) const override {
+    // TODO: use allocator that doesn't initialize memory?
+    std::string result(length, 0);
+    char* pos = result.data();
+    preadInternal(offset, length, pos);
+    return result;
+  }
+
+  uint64_t preadv(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>>& buffers) override {
+    throw std::runtime_error("failure in S3ReadFile::preadv.");
+    return -1;
+  }
+
+  uint64_t size() const override {
+    throw std::runtime_error("failure in S3ReadFile::size.");
+  }
+
+  uint64_t memoryUsage() const override {
+    throw std::runtime_error("failure in S3ReadFile::memoryUsage.");
+    return 0;
+  }
+
+  bool shouldCoalesce() const final {
+    return false;
+  }
+
+ private:
+  void preadInternal(uint64_t offset, uint64_t length, char* pos) const {
+    // Read the desired range of bytes
+    Aws::S3::Model::GetObjectRequest req;
+    Aws::S3::Model::GetObjectResult result;
+    auto first_sep = path_.find_first_of(kSep);
+    req.SetBucket(getAwsString(path_.substr(0, first_sep)));
+    req.SetKey(getAwsString(path_.substr(first_sep + 1)));
+    std::stringstream ss;
+    ss << "bytes=" << offset << "-" << offset + length - 1;
+    req.SetRange(getAwsString(ss.str()));
+    // TODO: Avoid copy below by using  req.SetResponseStreamFactory();
+    // Reference: ARROW-8692
+    auto outcome = client_->GetObject(req);
+    if (!outcome.IsSuccess()) {
+      throw std::runtime_error("failure in S3ReadFile::size preadInternal.");
```

Comment:
should we append some sort of error description to the exception being thrown? Also, if you throw using Velox's VELOX_CHECK() macros, it captures a lot of information (including the stack, file and lino, etc).

Reply:
Will improve this error check.